### PR TITLE
[IR] Handle `.__uniq.` in `getGlobalIndentifier` for local linkage objects

### PR DESCRIPTION
**Pull request by mtrofin**
The `.__uniq.` particle is used when the name of an object with local linkage incorporates a hash of its defining module - for example, if clang is passed `-funique-internal-linkage-names`. We don't need to add the module name in `getGlobalIndentifier` in this case.

We also don't need to apply the `.llvm.` particle when importing post-thinlink, if the function names were already uniq-ed - the former achieves the same goal the latter achieves.

This simplifies matters for carrying out information (like profiling data) between pre and post-thinlink: a function's ID stays the same from frontend through thinlink to the backend.

### CODE REVIEW

File: llvm/include/llvm/ProfileData/SampleProf.h [hunk 1]
```diff
@@ -1091,18 +1091,14 @@ class FunctionSamples {
     return getCanonicalFnName(F.getName(), Attr);
   }
 
-  /// Name suffixes which canonicalization should handle to avoid
```
Comment (minglotus-6) [inline llvm/include/llvm/ProfileData/SampleProf.h:1094]: pretty much a drive-by comment: preserve the comments (handle name to avoid profile mismatches) in a way

File: llvm/include/llvm/IR/ModuleSummaryIndex.h [hunk 2]
```diff
@@ -1743,8 +1743,10 @@ class ModuleSummaryIndex {
   }
 
   static std::string getGlobalNameForLocal(StringRef Name, StringRef Suffix) {
+    if (Name.contains(NameParticles::UniqSuffix))
```
Comment (minglotus-6) [inline llvm/include/llvm/IR/ModuleSummaryIndex.h:1746]: nit: add a comment to explain the reason behind this.

File: llvm/include/llvm/IR/ModuleSummaryIndex.h [hunk 3]
```diff
@@ -1743,8 +1743,10 @@ class ModuleSummaryIndex {
...
+      return std::string(Name);
     SmallString<256> NewName(Name);
-    NewName += ".llvm.";
+    NewName += NameParticles::LLVMSuffix;
     NewName += Suffix;
     return std::string(NewName);
```
Comment (ellishg) [inline llvm/include/llvm/IR/ModuleSummaryIndex.h:1751]: I'm not positive, but I wonder if `Twine` is faster.

```
return (Twine(Name) + NameParticles::LLVMSuffix + Suffix).str();
```

File: llvm/include/llvm/IR/ModuleSummaryIndex.h [hunk 4]
```diff
@@ -1743,8 +1743,10 @@ class ModuleSummaryIndex {
   }
 
   static std::string getGlobalNameForLocal(StringRef Name, StringRef Suffix) {
+    if (Name.contains(NameParticles::UniqSuffix))
+      return std::string(Name);
```
Comment (ellishg) [inline llvm/include/llvm/IR/ModuleSummaryIndex.h:1747]:
```suggestion
      return Name.str();
```

File: llvm/include/llvm/IR/ModuleSummaryIndex.h [hunk 5]
```diff
@@ -1754,7 +1756,8 @@ class ModuleSummaryIndex {
   /// because it is possible in certain clients (not clang at the moment) for
   /// two rounds of ThinLTO optimization and therefore promotion to occur.
   static StringRef getOriginalNameBeforePromote(StringRef Name) {
-    std::pair<StringRef, StringRef> Pair = Name.rsplit(".llvm.");
+    std::pair<StringRef, StringRef> Pair =
+        Name.rsplit(NameParticles::LLVMSuffix);
     return Pair.first;
```
Comment (ellishg) [inline llvm/include/llvm/IR/ModuleSummaryIndex.h:1761]:
```suggestion
  auto [Prefix, Suffix] = Name.rsplit(NameParticles::LLVMSuffix);
  return Prefix;
```

File: llvm/lib/IR/Globals.cpp [hunk 6]
```diff
@@ -159,11 +159,12 @@ std::string GlobalValue::getGlobalIdentifier(StringRef Name,
   Name.consume_front("\1");
 
   std::string GlobalName;
-  if (llvm::GlobalValue::isLocalLinkage(Linkage)) {
+  if (llvm::GlobalValue::isLocalLinkage(Linkage) &&
+      Name.find(NameParticles::UniqSuffix) == StringRef::npos) {
```
Comment (ellishg) [inline llvm/lib/IR/Globals.cpp:163]: Why not use `Name.contains(NameParticles::UniqSuffix)`?

File: llvm/lib/ProfileData/InstrProf.cpp [hunk 7]
```diff
@@ -599,10 +599,9 @@ StringRef InstrProfSymtab::getCanonicalName(StringRef PGOName) {
   // different modules and should be kept. This is the only suffix with the
   // pattern ".xxx" which is kept before matching, other suffixes similar as
   // ".llvm." will be stripped.
-  const std::string UniqSuffix = ".__uniq.";
-  size_t Pos = PGOName.find(UniqSuffix);
+  size_t Pos = PGOName.find(NameParticles::UniqSuffix);
```
Comment (ellishg) [inline llvm/lib/ProfileData/InstrProf.cpp:602]: I suspect this can actually be replaced with `getOriginalNameBeforePromote()`, but we might also need to strip out `PartSuffix` too. 

This reminds me of recent discussion on a different PR: https://github.com/llvm/llvm-project/pull/96268/files#r1651870215
CC @MaskRay

File: llvm/tools/llvm-profdata/llvm-profdata.cpp [hunk 8]
```diff
@@ -1142,7 +1142,7 @@ adjustInstrProfile(std::unique_ptr<WriterContext> &WC,
     }
 
     // This name should have a static linkage.
-    size_t PostfixPos = NewName.find(FunctionSamples::UniqSuffix);
+    size_t PostfixPos = NewName.find(NameParticles::UniqSuffix);
```
Comment (ellishg) [inline llvm/tools/llvm-profdata/llvm-profdata.cpp:1145]: Can we use `StringRef::contains()` here too? Same with above, but I suspect `toString()` returns `std::string`.

**Comment by mtrofin** [comment]
Abandoning this - this leaves unsolved the case where externs are internalized post-link. I have an alternative proposal which addresses that and also sets us up to do away with the various name -> GUID gymnastics we do in InstrProf and elsewhere. Will send a RFC shortly, want to make sure "it holds".

**Comment by ellishg** [comment]
It might be worthwhile to create a simple NFC PR that simply creates a global `NameParticles` struct and removes the string constants like `".llvm."`. This has a nice benefit of making these searchable in the codebase.

**Comment by mtrofin** [comment]
> It might be worthwhile to create a simple NFC PR that simply creates a global `NameParticles` struct and removes the string constants like `".llvm."`. This has a nice benefit of making these searchable in the codebase.

I agree, I should be able to split that off this quickly
